Mathlib/Tactic/TFAE.lean

```diff
@@ -15,48 +16,111 @@ This file provides the tactics `tfae_have` and `tfae_finish` for proving goals o
 `TFAE [P₁, P₂, ...]`.
 -/
 
-open List Lean Meta Expr Elab.Term Elab.Tactic Mathlib.Tactic Qq
-
 namespace Mathlib.Tactic.TFAE
 
-/-- An arrow of the form `←`, `→`, or `↔`. -/
-syntax impArrow := " → " <|> " ↔ " <|> " ← "
+/-! # Parsing and syntax
+
+We implement `tfae_have` in terms of a syntactic `have`. To support as much of the same syntax as
+possible, we recreate the parsers for `have`, except with the changes necessary for `tfae_have`.
+-/
+
+open Lean.Parser Term
+
+namespace Parser
+
+/- An arrow of the form `←`, `→`, or `↔`. -/
+private def impTo : Parser := leading_parser unicodeSymbol " → " " -> "
+private def impFrom : Parser := leading_parser unicodeSymbol " ← " " <- "
+private def impIff : Parser := leading_parser unicodeSymbol " ↔ " " <-> "
+private def impArrow : Parser := leading_parser impTo <|> impFrom <|> impIff
```

Comment:
What's the difference between this and the shorter:
```suggestion
syntax impArrow := unicodeSymbol " → " " -> " <|> unicodeSymbol " ← " " <- " <|> unicodeSymbol " ↔ " " <-> "
```

Reply:
The (unfortunate) reasons are that
* `unicodeSymbol` doesn't work in `syntax` (there's no parser alias that works in `stx` syntax)
* if we try inlining all three but leave it as a `: Parser`, then the syntax quotation matches `` `(impArrow| → ) `` below stop working

If we made a parser alias for `unicodeSymbol`, then all of the parsers following this would also have to be changed to `syntax` (as `ParserDescr`s created by `syntax ... := ...` are not allowed to in `Parser`s)—this might be a good thing, though! However, from what I can see, parser aliases like `many(p)` can't take non-parser arguments (such as strings), so it might be difficult to get `unicodeSymbol("u","a") : stx` to work. (Maybe there's a way to work around this somehow, by extracting the strings from the parsers `"someSymbol"` and failing otherwise?)

Reply:
I don't understand at all why things work the way it works, for example the following works too:

```
private def impTo : Parser := leading_parser unicodeSymbol " → " " -> "
private def impArrow : Parser := leading_parser
  impTo <|>
  (leading_parser unicodeSymbol " ← " " <- ") <|>
  ( unicodeSymbol " ↔ " " <-> ")
```

but any other combination of (not) using `leading_parser` seems to fail.

I suggest that @thorimur 's version seems reasonable to keep as-is.